#CRISPResso_BE_analysis_functs.R
# conda_environment: crispresso_downstream_env
# last modified: 2020_07_21 Anne Shen
# For use with CRISPResso version 2.0.37
#
###  Dependencies:
# library(tidyselect)
# library(tidyverse)
# library(RColorBrewer)
# library(grid)
# options(scipen=999) #turn off scientific notation
#
### Functions:
# get_BE_summary_tb()




###### get_BE_summary_tb() #############################################################################
# For CRISPResso2 outputs from base editor samples. Assumes the quantification window
# covers the entire spacer sequence.
# This function takes a list of "_all_[percent_cutoff].csv" summary allele tables and a data column 
# identifer and filters out likely sequencing errors (keeps C>N substitions in spacer bps 
# 3-10 and indels in bp 17-18 spacer sequence).
#
# ARGUMENTS:  summary_file = list of summary files ("_all_[percent_cutoff].csv")
#             dataID = string idenitfier in the sample names that identifies a column as a column of 
#                      allele frequencies
#             ref_nucleotide = nucleotide targeted by BE conversion (ex. C)
#             target_nucleotides = nucleotide(s) generated by BE (no separators, ex. ATCG)
# OUTPUT: none (directly saves summary tables to .csv files in current directory)
get_BE_summary_tb <- function(summary_files, dataID, percent_freq_cutoff, ref_nucleotide, target_nucleotides){
  
  # tester 
  # summary_files <- list.files(path = ".", pattern = "collapse")
  # dataID <- 1620
  # percent_freq_cutoff <- 0 
  # ref_nucleotide <- "C"
  # target_nucleotides <- "T"
  
  #generate analysis log
  #sink("BaseEdit_summary_log.txt", append=FALSE, split=TRUE)
  
  cat("BaseEdit_summary_log\n",
      paste(Sys.time(), "\n", sep = ""),
      paste(getwd(), "\n", sep = ""),
      "dataID: ", dataID, "\n",
      "percent_freq_cutoff: ", percent_freq_cutoff, "\n",
      "ref_nucleotide: ", ref_nucleotide, "\n",
      "target_nucleotide(s): ", target_nucleotides, "\n",
      "\n",
      "Summarizing...\n\n")
  
  #generate regular expressions for base edit searches (ex. "C[3-9]{1}[ATCG]")
  bp_3_9_target_regexpr <- paste(ref_nucleotide, "[3-9]{1}[", target_nucleotides, "]", sep = "")
  bp_10_target_regexpr <- paste(ref_nucleotide, "10[", target_nucleotides, "]", sep = "")
  
  for(n in seq(1,length(summary_files))){

    #get csv_prefix for saving file
    csv_prefix <- sub(paste("_collapsed_", percent_freq_cutoff, ".csv", sep = ""), "", summary_files[n])
    
    #cat in log file
    cat(csv_prefix, "\n")
    
    #read in summary file
    summary_table <- read.csv(summary_files[n], check.names = FALSE, stringsAsFactors = FALSE)
    
    #initialize edits tracking vector
    edited_alleles <- c()
    
    #### Filter for "correct" base edits (only) and collapse by substitution 
    real_be_only_tb <- summary_table %>% 
      filter((grepl(bp_3_9_target_regexpr, summary_table$indel) | 
                grepl(bp_10_target_regexpr, summary_table$indel)) & 
               !grepl("[+,-]", summary_table$indel))
    
    #extract the "real" base edits and format the indel column
    real_be_tb <- real_be_only_tb %>%
      mutate(be_3_9 = gsub("character\\(0\\)", "",
                           regmatches(real_be_only_tb$indel, 
                                 gregexpr(bp_3_9_target_regexpr, real_be_only_tb$indel))),
             be_10 = gsub("character\\(0\\)", "", 
                          regmatches(real_be_only_tb$indel, 
                                     gregexpr(bp_10_target_regexpr, real_be_only_tb$indel)))) %>%
      transform(be_3_9 = gsub("c\\(\"", "", 
                              gsub("\", \"", " ",
                                   gsub("\"\\)", "", be_3_9)))) %>%
      unite(indel, c("be_3_9", "be_10"), sep = " ") %>%
      transform(indel = trimws(indel, which = "both"))
    
    #collapse alleles by indel
    if(nrow(real_be_tb) > 0){
      real_be_tb <- collapse_duplicate_indels(real_be_tb, dataID)[1:ncol(real_be_tb)]
      
      #keep track of "edited" indels to filter out when extracting "unedited" alleles
      edited_alleles <- real_be_only_tb$indel
    }
    
    
    #### Filter for "real" indels (only) within entire spacer sequence and collapse by indel
    indel_only_tb <- summary_table %>% 
      filter(grepl("[+,-]", summary_table$indel) & 
               !(grepl(bp_3_9_target_regexpr, summary_table$indel) &
                   !grepl(bp_10_target_regexpr, summary_table$indel)))
    
    #extract the "real" indels and format the indel column
    indel_only_tb <- indel_only_tb %>%
      transform(indel = regmatches(indel_only_tb$indel, 
                                   regexpr("[+,-][0-9]{1,} ([[:print:]]{1,})", 
                                           indel_only_tb$indel))) %>%
      mutate(indel_range = trimws(gsub("(?<=[0-9])-", "_",
                                 gsub("[[:alpha:]]{1, }", "", 
                                      gsub("[-,+]{1}[0-9]{1,} \\(", "",
                                           gsub("\\)", "", indel, perl = TRUE))), 
                                perl = TRUE))) %>%
      separate(col = indel_range, into = c("indel_start", "indel_end"), sep = "_", remove = FALSE) %>%
      transform(indel_end = ifelse(is.na(indel_end), 0, indel_end)) %>%
      mutate(real_indel = ifelse((as.numeric(indel_start) <= 18 & as.numeric(indel_end) >= 17), TRUE, 
                                 ifelse((as.numeric(indel_start) == 17 | as.numeric(indel_start) == 18), TRUE, FALSE)))
    
    real_indel_tb <- indel_only_tb %>%
      filter(real_indel) %>%
      select(-c("indel_range", "indel_start", "indel_end", "real_indel"))
    
    #collapse alleles by indel
    if(nrow(real_indel_tb) > 0){
      real_indel_tb <- collapse_duplicate_indels(real_indel_tb, dataID)[1:ncol(real_indel_tb)]

      #keep track of "edited" indels to filter out when extracting "unedited" alleles
      edited_alleles <- c(edited_alleles, indel_only_tb$indel[which(indel_only_tb$real_indel)])
    }
    
    
    #### Filter for "real" indels and BE within entire spacer sequence and collapse by indel
    real_be_and_indel_tb <- summary_table %>%
      filter(grepl("[+,-]", summary_table$indel) &
               (grepl(bp_3_9_target_regexpr, summary_table$indel) |
                  grepl(bp_10_target_regexpr, summary_table$indel)))

    #keep track of "edited" indels to filter out when extracting "unedited" alleles
    edited_alleles <- c(edited_alleles, real_be_and_indel_tb$indel)

    #extract the "real" edits and format the indel column
    real_be_and_indel_tb <- real_be_and_indel_tb %>%
      mutate(be_3_9 = regmatches(real_be_and_indel_tb$indel,
                                 gregexpr(bp_3_9_target_regexpr, real_be_and_indel_tb$indel)),
             be_10 = gsub("character\\(0\\)", "",
                          regmatches(real_be_and_indel_tb$indel,
                                     gregexpr(bp_10_target_regexpr, real_be_and_indel_tb$indel))),
             in_del =  regmatches(real_be_and_indel_tb$indel,
                                  regexpr("[+,-][0-9]{1,} ([[:print:]]{1,})",
                                          real_be_and_indel_tb$indel))) %>%
      transform(be_3_9 = gsub("c\\(\"", "",
                              gsub("\", \"", " ",
                                   gsub("\"\\)", "", be_3_9)))) %>%
      unite(be_total, c("be_3_9", "be_10"), sep = " ") %>%
      transform(be_total = trimws(be_total, which = "both")) %>%
      mutate(indel_range = trimws(gsub("(?<=[0-9])-", "_",
                                       gsub("[[:alpha:]]{1, }", "", 
                                            gsub("[-,+]{1}[0-9]{1,} \\(", "",
                                                 gsub("\\)", "", in_del, perl = TRUE))), 
                                       perl = TRUE))) %>%
      separate(col = indel_range, into = c("indel_start", "indel_end"), sep = "_", remove = FALSE) %>%
      transform(indel_end = ifelse(is.na(indel_end), 0, indel_end)) %>%
      mutate(real_indel = ifelse((as.numeric(indel_start) <= 18 & as.numeric(indel_end) >= 17), TRUE, 
                                 ifelse((as.numeric(indel_start) == 17 | as.numeric(indel_start) == 18), 
                                        TRUE, FALSE))) %>%
      transform(in_del = ifelse(real_indel, in_del, "")) %>%
      unite(indel, c("be_total", "in_del"), sep = ", ") %>%
      transform(indel = gsub(", $", "", indel)) %>%
      select(-c("indel_range", "indel_start", "indel_end", "real_indel"))
  
    #collapse alleles by indel
    if(nrow(real_be_and_indel_tb) > 0){
      real_be_and_indel_tb <- collapse_duplicate_indels(real_be_and_indel_tb, dataID)[1:ncol(real_be_and_indel_tb)]
    }

    #bind tables of "true" edits together
    real_edit_tb <- rbind(rbind(real_be_tb, real_indel_tb), real_be_and_indel_tb) %>%
      transform( indel = trimws( indel, which = "both"))
    
    #collapse alleles by indel
    if(nrow(real_edit_tb) > 0){
      real_edit_tb <- collapse_duplicate_indels(real_edit_tb, dataID)[1:ncol(real_edit_tb)]
    }

    names(real_edit_tb) <- gsub("^X", "", names(real_edit_tb))


    #get table of "false" edits and unedited percentages and sum the percentages to generate
    # total unedited percentage
    unedited_tb <- summary_table[! (summary_table$indel %in% edited_alleles), ]

    unedited_tb[1,vars_select(names(unedited_tb),contains(dataID))] <- unedited_tb %>%
      select(vars_select(names(unedited_tb),contains(dataID))) %>%
      colSums(na.rm = TRUE)

    #bind unedited and edited tables together
    real_alleles_tb <- rbind(unedited_tb[1,], real_edit_tb)
    real_alleles_tb <- adjust_percent_frequency(real_alleles_tb, dataID)

    #save as .csv
    BE_conversion <- paste("Cto", gsub("[[:punct:]]", "", target_nucleotides), sep = "")
    save_to_csv(real_alleles_tb, ".", paste(csv_prefix,"BE_summary", BE_conversion, sep = "_"))
  }
  
  #end summary run log
  cat("\n")
  #sink()
}
